adds support for installer types inno, nsis, wise and installshield to windows_package resource

mwrock opened the pull request:
This initial commit is a direct port from the Windows cookbook. There will be some implementation changes particularly around installer detection and idempotency but this is the basic structure I'm proposing:

* Keep the current `windows_package` provider with its `MSI` as an inner class
* Add a new inner class `Exe` to encapsulate the other installers

I don't think we need separate classes for each installer type because the implementation is almost identical with the only difference being in the installer flags. They are different enough from MSI to warant their own class.

Note that the private methods of the `Exe` class may be undergoing changes so hold off on reviewing the specifics of that code. Also expect the `installer_type` detection to change because we cant be opening up hundred MB files in memory all at once.

mwrock commented:
cc @lamont-granquist @chef/client-windows

lamont-granquist commented:
Yeah seems fine to get this done.  There's probably two ways to do this, and this is going to be the way that involves a lot fewer objects.

btm commented on lib/chef/provider/package/windows.rb:
```diff
@@ -64,16 +62,36 @@ def package_provider
+                  if basename == 'setup.exe'
+                    :installshield
+                  else
+                    fail Chef::Exceptions::AttributeNotFound, "Installer type for Windows Package '#{@new_resource.name}' not specified and cannot be determined from file extension '#{file_extension}'"
```
I wouldn't use AttributeNotFound here, as it's used for something else specific, but I seem to be in an echo chamber so maybe it's not important.

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,159 @@
+#
+# Author:: Matt Wrock <matt@mattwrock.com>
+# Copyright:: Copyright (c) 2015 Chef Software, Inc.
```
Given this is mostly copy-paste from the original provider, you add another author line for `Author:: Seth Chisamore <schisamo@chef.io>` and add 2011 to the copyright?

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,159 @@
+# TODO: Allow @new_resource.source to be a Product Code as a GUID for uninstall / network install
+
+require 'chef/mixin/shell_out'
+require 'Win32API' if Chef::Platform.windows?
```
What's this for?

mwrock replied:
Part of the copy/paste from the cookbook. I'm guessing its for the registry stuff?? Will remove if unnecessary or not used.

btm replied:
tests still passed without it. I can't see any use of it in core, I do see it in the cookbook.

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,159 @@
+          end
+
+          def install_package(name, version)
+            Chef::Log.debug("#{@new_resource} installing Nullsoft package '#{@new_resource.source}'")
```
replace nullsoft with @new_resource.installer_type

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,159 @@
+                                         nil
+                                       end
+                    if display_name
+                      packages[display_name] = { name: display_name,
```
note that the usage of display_name as the key here prevents handle two versions of a package with the same name. this is an existing bug, but consider fixing it if doing so would break something.

https://github.com/chef-cookbooks/windows/issues/192

mwrock replied:
yeah I think this will be addressed in the next phase that polishes idempotency.

lamont-granquist commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,159 @@
+                                     k['DisplayName']
+                                   rescue
+                                     nil
+                                   end
```
i think you can use `display_name = k['DisplayName'] rescue nil`

mwrock replied:
I think we are gonna take a different strategy here. As @btm points out, there can be multiple versions of a "package" installed that share the same display name. We should really be using the leaf key as a unique identifier. So our "current_version" would possibly be a collection of versions of all installed packages with the same display name. We would converge if:
* the new_version is different from all installed version
* there are no installed versions

Further, currently a new_version can only be specified via an attribute. I believe we can improve on this by looking at the file properties (this is different from the properties exposed via an MSI). I know that `inno` and null soft (probably wise and install shield too) expose file properties that line up with the installed version. We can make a "best effort" to grab those and avoid a converge if they match with any installed version.

lamont-granquist replied:
That was more just a style feedback...

mwrock replied:
ah. ok

mwrock commented:
This is still WIP. This makes the following changes to `Exe` based packages:

* We track all versions installed and key by version (we only store for the package being installed/uninstalled)
* We install if either no version is given or the given version is not installed

When uninstalling:
* If a version is given, uninstall that version and raise if its not installed
* If no version is given, uninstall package if a single version is installed
* If no version is given and multiple versions are installed, raise

btm commented on lib/chef/provider/package/windows.rb:
```diff
@@ -56,7 +50,11 @@ def package_provider
+              require 'chef/provider/package/windows/exe'
+              Chef::Provider::Package::Windows::Exe.new(resource_for_provider, installer_type)
             else
               raise "Unable to find a Chef::Provider::Package::Windows provider for installer_type '#{installer_type}'"
```
I think you meant to raise `CannotDetermineWindowsInstallerType` here.

mwrock commented:
Still WIP. Currently functional and unit tests are green, I rebased to current master, and we have full functional parity with windows cookbook. Here is what needs to happen before merging:

* Some basic refactoring. Things are sloppy in a couple places and some unit tests are "awkward".
* Need to add wise and install shield to the functional test suite. Just need to find sample installers.
* Add logic to installer_type discovery that can infer installer type from registry information in the case on uninstalls where the installer is not available.

mwrock commented:
freshly rebased and squashed for your reviewing pleasure. Removed WIP flag. The last two bullets from the comment above still stand but I believe this can be "shipped" as is.

mwrock commented:
Here are a couple design decisions I'd like to lay out for discussion:
1. If uninstalling a package and there are multiple versions and no version is specified to remove, we throw an error basically saying we don't know which version to remove. This seems better than removing all versions without confirmation. I also see this as a very rare edge case. Its highly unlikely to have two installs with the exact same display name (the best key unfortunately for non MSIs)
2. If uninstalling a package and the resource points to an installer but that installer has a different version than the one installed, the resource will not converge the `:remove` action. I think this is "technically" correct and perhaps most uninstall actions will not include the package source (a local file path or HTTP url) in the resource. However, I could see this potentially being an issue with self updating software where it is likely that the version installed has surpassed the media used to run the install.

btm commented:
1. I'm concerned about raising when there are multiple packages to uninstall because it makes it hard / impossible to say "make sure this package isn't installed"

I've confirmed this manually too, but the gem package provider (and chef_gem) remove all versions if you haven't specified a version.

https://github.com/chef/chef/blob/98d553ced4bde95210fac82006c9be24ddb8f817/lib/chef/provider/package/rubygems.rb#L565

btm commented:
@chef/client-windows ready for review. we'd like to merge this within 24 hours to get it into the 12.6 release.

btm commented on lib/chef/provider/package/windows.rb:
```diff
@@ -56,24 +50,62 @@ def package_provider
             when :msi
+              require 'chef/provider/package/windows/msi'
               Chef::Provider::Package::Windows::MSI.new(resource_for_provider)
             else
```
what happens if I pass an msu or some other wierd non-exe thing?

mwrock replied:
today they would have to be treated as a custom type. my guess is that `MSU`s have the same "product" semantics as MSIs. I'll test and if so we could juse the  MSI class for MSUs.

btm commented on lib/chef/provider/package/windows.rb:
```diff
@@ -56,24 +50,62 @@ def package_provider
+                 end
+
+                # if file is named 'setup.exe' assume installshield
+                if basename == 'setup.exe'
```
if we're going to presume installshield anyway, can we avoid the file I/O against setup.exe?

mwrock replied:
question is if the other installers allow the installer file to be called `setup.exe` and if so, then we should inspect the file to allow them to catch it. I'm not certain but I'm assuming it is possible. Need to investigate.

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,177 @@
+# limitations under the License.
+#
+
+# TODO: Allow @new_resource.source to be a Product Code as a GUID for uninstall / network install
```
are GUID's still applicable to non-msi packages?

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,177 @@
+
+          def uninstall_command(uninstall_string)
+            uninstall_string.delete!('"')
+            if uninstall_string.start_with?("MsiExec.exe ")
```
where does `MsiExec.exe ` come from? is it guaranteed to always be cased this way?

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,177 @@
+              else
+                version_info = Chef::ReservedNames::Win32::File.version_info(@new_resource.source)
+                file_version = version_info.FileVersion || version_info.ProductVersion
+                file_version == '' ? nil : file_version
```
`file_version.empty? ? nil : file_version` ?

the double question marks look wierd, but ` == ''` looks weird to me. *shrug*

mwrock replied:
generally I like `empty?` but `file_version` could be `nil` here

btm commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,177 @@
+            end
+          end
+
+          def extract_installed_packages_from_key(hkey = ::Win32::Registry::HKEY_LOCAL_MACHINE, desired = ::Win32::Registry::Constants::KEY_READ)
```
`require 'win32/registry'` somewhere in this file?

btm commented on RELEASE_NOTES.md:
```diff
@@ -41,3 +41,8 @@ also array of versions and array of sources).
+
+## `windows_package` now supports non-`MSI` based installers
+
+Today you can install `MSI`s using the `windows_package` resource. However, you have had to use the windows cookbook in order to install non `MSI` based installer packages such as Nullsoft, Inno Setup, Installshield and other `EXE` based installers. We have moved and slightly improved the windows cookbook resource into the core chef client. This means you can now run most windows installer types  ...
```
we should note in DOC_CHANGES.md that the possible installer_types change.

https://docs.chef.io/resource_windows_package.html

btm commented on lib/chef/win32/api/file.rb:
```diff
@@ -565,6 +600,21 @@ def retrieve_file_info(file_name)
+          file_name = encode_path(file_name)
+          file_size = GetFileVersionInfoSizeW(file_name, "")
+          if file_size == 0
+            Chef::ReservedNames::Win32::Error.raise!
```
I wonder if it makes sense to call `Chef::ReservedNames::Win32::Error.get_last_error` here and print that in the exception. I thought we did that but apparently the only place we reference it is in some of the file security code.

mwrock replied:
`Chef::ReservedNames::Win32::Error.raise!` will do this already

btm commented:
can you add a functional test for the File::VersionInfo class?

mwrock commented:
I have finished up addressing review comments as well as adding support for removing packages when the original installer is not on disk and removing all versions when there are multipe versions and no version is specified.

There was a comment about supporting `:MSU` and I think that should be added but in a separate PR since it is not breaking functionality.

The only thing remaining for this PR is the doc changes which I will have shortly.

mwrock commented:
doc changes are in and commits resquashed

jaym commented on lib/chef/provider/package/windows.rb:
```diff
@@ -106,15 +142,59 @@ def new_version_array
+          @uninstall_registry_entries ||= Chef::Provider::Package::Windows::RegistryUninstallEntry.find_entries(new_resource.name)
+        end
+
+        def inferred_registry_type
```
does this work with multipackage things?

mwrock replied:
yes. it will return an entry for each

mwrock replied:
let me clarify, "multipackage" for `windows_package` is multiple versions for the same package. `windows_package` does not support installing multiple packages.

jaym commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,125 @@
+#
```
going to make a general style comment on this file:
you should avoid accessing variables throught `@`. The reason is because if you mistype something, ruby will happily continue along.

mwrock replied:
seems like a good rule.

chefsalim commented on DOC_CHANGES.md:
```diff
@@ -78,3 +78,42 @@ either reboot immediately (:reboot_now) or queue a reboot (:request_reboot).  Th
+
+`windows_package` now supports more than just `MSI`. Most common windows installer types are supported including inno setup, nullsoft, wise and installshield. The new allowed `installer_type` values are: `inno`, `nsis`, `wise`, `installshield`, `custom`, and `msi`.
+
+Also, while being able to download remote installers from a `HTTP` resource is not new, it looks as though the top of the docs page is incorrect stating that only local installers can be used as a source.
```
Slightly confusing - can we just fix the doc instead of calling out they are incorrect?

mwrock replied:
I might misunderstand this file (first time editing it), but i thought thats whthis is for to alert the docs witers what needs to be changed in the official docs.

chefsalim replied:
Hmm, I have assumed in my earlier edits that this file is for letting people know what docs have changed.

chefsalim commented on DOC_CHANGES.md:
```diff
@@ -78,3 +78,42 @@ either reboot immediately (:reboot_now) or queue a reboot (:request_reboot).  Th
+
+### `windows_package` resource
+
+`windows_package` now supports more than just `MSI`. Most common windows installer types are supported including inno setup, nullsoft, wise and installshield. The new allowed `installer_type` values are: `inno`, `nsis`, `wise`, `installshield`, `custom`, and `msi`.
```
Capitalize the installer names (similar to below)

jaym commented on lib/chef/provider/package/windows/registry_uninstall_entry.rb:
```diff
@@ -0,0 +1,78 @@
+                    end
+                  end
+                end
+              rescue ::Win32::Registry::Error
```
Log message?

jaym commented on lib/chef/provider/package/windows/registry_uninstall_entry.rb:
```diff
@@ -0,0 +1,78 @@
+                      if display_name == package_name
+                        entries.push(RegistryUninstallEntry.new(hkey, key, k))
+                      end
+                    rescue ::Win32::Registry::Error
```
Log message?

jaym commented on lib/chef/provider/package/windows/registry_uninstall_entry.rb:
```diff
@@ -0,0 +1,78 @@
+            @key = key
+            @data = registry_data
+            @display_name = registry_data['DisplayName']
+            @display_version = registry_data['DisplayVersion'] rescue nil
```
I'm not a fan of rescuing all the exceptions. I'm guessing in this case, you're expecting a registry error if the thing you're looking for cant be found? I think you should call a function like `maybe_registry_value` that only catches that error. I would still log to debug if you cant find it.

mwrock replied:
i like it

jaym commented on lib/chef/resource/windows_package.rb:
```diff
@@ -33,7 +33,9 @@ class WindowsPackage < Chef::Resource::Package
       def initialize(name, run_context=nil)
         super
-        @source ||= source(@package_name)
+        if uri_scheme?(@package_name) || ::File.exists?(canonical_path(@package_name))
```
Something seems wrong here. What if I have a recipe that does a `cookbook_file` to for getting the package, and then a `windows_package`? I think the file will be neither a uri or exist, and `@source` will not be set?

jaym commented on lib/chef/win32/api/file.rb:
```diff
@@ -565,6 +600,21 @@ def retrieve_file_info(file_name)
 
+        def retrieve_file_version_info(file_name)
+          file_name = encode_path(file_name)
+          file_size = GetFileVersionInfoSizeW(file_name, "")
```
The second parameter here needs to be a pointer to a DWORD according to https://msdn.microsoft.com/en-us/library/windows/desktop/ms647005%28v=vs.85%29.aspx

mwrock replied:
yeah. "A pointer to a variable that the function sets to zero." We don't do anything with it. Is there a potential issue with passing an empty string instead of creating a pointer variable?

jaym replied:
well, an empty string could be 2 bytes

mwrock replied:
ah right. makes sense

jaym commented on lib/chef/win32/api/file.rb:
```diff
@@ -565,6 +600,21 @@ def retrieve_file_info(file_name)
+            Chef::ReservedNames::Win32::Error.raise!
+          end
+          
+          version_info = ' '*file_size
```
This makes me uneasy. While it may happen to work, there are better ways to do this. We should really be allocating a buffer and passing that through.

jaym commented on lib/chef/win32/file/version_info.rb:
```diff
@@ -0,0 +1,93 @@
+        def translation
+          @translation ||= begin
+            info_ptr = FFI::MemoryPointer.new(:pointer)
+            unless VerQueryValueW(@file_version_info, "\\VarFileInfo\\Translation".to_wstring, info_ptr, "")
```
this function should be wrapped in a function inside our win32 code that abstracts out all the ffi specific things. It should take in pure ruby types

jaym commented on spec/functional/win32/version_info_spec.rb:
```diff
@@ -0,0 +1,50 @@
+  let(:os_version) do
+    wmi = WmiLite::Wmi.new
+    os_info = wmi.first_of('Win32_OperatingSystem')
+    os_info['version']
```
I don't know how fast this call is, but if it's not, it might be something to cache (`let!` i think). If it is fast, ignore this comment

chefsalim commented on lib/chef/provider/package/windows/registry_uninstall_entry.rb:
```diff
@@ -0,0 +1,78 @@
+          def self.find_entries(package_name)
+            entries = []
+            [
+              [::Win32::Registry::HKEY_LOCAL_MACHINE, (::Win32::Registry::Constants::KEY_READ | 0x0100)],
```
I don't recall now what happens if you pass in 0x0100 to a pure 32-bit Windows .. would be good to test if you haven't already

chefsalim commented on lib/chef/provider/package/windows/exe.rb:
```diff
@@ -0,0 +1,125 @@
+        class Exe
+          include Chef::Mixin::ShellOut
+
+          def initialize(resource, installer_type, uninstall_entries)
```
From an encapsulation standpoint, should we pass in the installer type and uninstall_entries, or should the provider determine them on it's own?

mwrock replied:
yeah I considered doing something like that where the MSI and EXE classes register themselves and then the provider interates each registered class and "asks" them if they support the file. I was going to leave that for another pr.

chefsalim commented on lib/chef/provider/package/windows/registry_uninstall_entry.rb:
```diff
@@ -0,0 +1,78 @@
+                ::Win32::Registry.open(hkey[0], UNINSTALL_SUBKEY, desired) do |reg|
+                  reg.each_key do |key, _wtime|
+                    begin
+                      k = reg.open(key, desired)
```
Minor nitpick... might be good to change 'k' to a bit more descriptive name.

mwrock replied:
yup. I blindly copied some of this over from the windows cookbook a little too blindly

chefsalim commented on RELEASE_NOTES.md:
```diff
@@ -52,6 +52,10 @@ Please see the following for more details : https://docs.chef.io/release/12-6/re
 
 This is the first release where we are rolling out a MSI package for Windows that significantly improves the installation time. In a nutshell, the new approach is to deploy and extract a zipped package rather than individually tracking every file as a MSI component. Please note that the first  upgrade (ie, an older version of Chef client is already on the machine) may not exhibit the full extent  ...
 
+## `windows_package` now supports non-`MSI` based installers
```
Could you make a small change and add a " (Windows)" to the end of the title to be consistent with the other notes.

And sorry for the re-base .. hope it's not too painful.

chefsalim commented:
LGTM..don't have anything other than the minor nitpicks.. :+1:

jaym commented:
:+1:
